### ISSUE TITLE
Check if use statement already defined

### ISSUE THREAD

**Issue by orklah** [rectorphp/rector#977]
Hi,

After running twig240 level, I had the following suggestion:

```rector
    ---------- begin diff ----------
--- Original
+++ New
@@ -5,9 +5,9 @@

 use Twig\Extension\AbstractExtension;
 use Twig\TwigFilter;
-use Twig_Filter;
+use \Twig\TwigFilter;

-class ShortUrlExtension extends AbstractExtension
+class ShortUrlExtension extends \Twig\Extension\AbstractExtension
 {
     /**
      * @return Twig_Filter[]
@@ -15,7 +15,7 @@
     public function getFilters(): array
     {
         return [
-            new TwigFilter('shortUrl', [$this, 'shortUrl']),
+            new \Twig\TwigFilter('shortUrl', [$this, 'shortUrl']),
         ];
     }
    ----------- end diff -----------

Applied rectors:

 * Rector\Rector\Class_\ClassReplacerRector

```

I applied the suggestion and it added "\Twig\TwigFilter" when "Twig\TwigFilter" was already there.

It produce a Fatal when executed: 

> FatalErrorException
> Compile Error: Cannot use Twig\TwigFilter as TwigFilter because the name is already in use

It's a small issue and easy to fix manually but it would be nice if it was solved from the beggining.

Thanks!

**Comment by orklah** [comment]
I just noticed two other issues on the same file:
- I don't understand the reason behind replacing AbstractExtension by \Twig\Extension\AbstractExtension as I already had the use statement defined.
- The rector didn't change the @return Twig_Filter on the phpdoc

Thanks

**Comment by TomasVotruba** [comment]
This should be resolved by `--with-style` or your custom coding standard.

**Comment by TomasVotruba** [comment]
> * The rector didn't change the @return Twig_Filter on the phpdoc

Valid :+1:

**Comment by TomasVotruba** [comment]
Oh, I see it re-replaces existing class. That's a bug

**Comment by orklah** [comment]
> > * The rector didn't change the @return Twig_Filter on the phpdoc
> 
> Do you think Rector should take care of that?

I thought that was supposed to be the case and that was a bug. If it's possible, it could be nice to do it  automatically. It's a pity we have to go modify the phpdoc when rector was confident enough to make the change in the code by itself :)

**Comment by TomasVotruba** [comment]
> I thought that was supposed to be the case and that was a bug. If it's possible, it could be nice to do it automatically. It's a pity we have to go modify the phpdoc when rector was confident enough to make the change in the code by itself :)

The problem is AST doesn't work with docs. I'll check how difficult this change would be.

> I don't understand the reason behind replacing AbstractExtension by \Twig\Extension\AbstractExtension as I already had the use statement defined

I'm trying to reproduce this, but there is no parent class replacement: https://github.com/rectorphp/rector/pull/1036/commits/c2014eeb40595f136e87841df8531126116dadd1

**Comment by orklah** [comment]
> > I don't understand the reason behind replacing AbstractExtension by \Twig\Extension\AbstractExtension as I already had the use statement defined
> 
> I'm trying to reproduce this, but there is no parent class replacement. See this commit [c2014ee](https://github.com/rectorphp/rector/commit/c2014eeb40595f136e87841df8531126116dadd1)

I can still reproduce on v0.3.41:

```
    ---------- begin diff ----------
--- Original
+++ New
@@ -6,7 +6,7 @@
 use Twig\Extension\AbstractExtension;
 use Twig\TwigFilter;

-final class DirnameExtension extends AbstractExtension
+final class DirnameExtension extends \Twig\Extension\AbstractExtension
 {
     /**
      * @return TwigFilter[]
@@ -14,7 +14,7 @@
     public function getFilters(): array
     {
         return [
-            new TwigFilter('dirname', [$this, 'dirname']),
+            new \Twig\TwigFilter('dirname', [$this, 'dirname']),
         ];
     }
    ----------- end diff -----------

```

here is the code of the class:
```php
<?php
declare(strict_types=1);

namespace App\Twig;

use Twig\Extension\AbstractExtension;
use Twig\TwigFilter;

final class DirnameExtension extends AbstractExtension
{
    /**
     * @return TwigFilter[]
     */
    public function getFilters(): array
    {
        return [
            new TwigFilter('dirname', [$this, 'dirname']),
        ];
    }

    /**
     * @param string $path
     * @return string
     */
    public function dirname(string $path): string
    {
        return \dirname($path);
    }
}
```

**Comment by TomasVotruba** [comment]
Could you remove the filter class? I wonder if this affects the change:

```diff
 public function getFilters(): array
 {
     return [
-        new TwigFilter('dirname', [$this, 'dirname']),
     ];
 }
```

Then run Rector command. Also, what exact `vendor/bin/rector` command do you run?

**Comment by orklah** [comment]
I reproduce it with as little as 

```php
<?php

namespace App\Twig;

use Twig\Extension\AbstractExtension;

final class DirnameExtension extends AbstractExtension
{

}
```
here is the command:
vendor\bin\rector -d process src --level="twig240"

**Comment by TomasVotruba** [comment]
I see. Thanks for minimal example :+1: 

Last idea I have is operating system. What do you use?

**Comment by orklah** [comment]
I'm on windows. You think this has something to do with it?

**Comment by TomasVotruba** [comment]
I though so. It's usually the last weak spot, but filesystem differences + `\` vs `/` can break lot of code.

**Comment by TomasVotruba** [comment]
Could you check fis this line is called?

https://github.com/rectorphp/rector/blob/6ba0b32f60f3326035f071b7e2a81d34d36ad431/src/Rector/Class_/ClassReplacerRector.php#L90

That's the one that change class names.

**Comment by orklah** [comment]
> Could you check fis this line is called?
> 
> [rector/src/Rector/Class_/ClassReplacerRector.php](https://github.com/rectorphp/rector/blob/6ba0b32f60f3326035f071b7e2a81d34d36ad431/src/Rector/Class_/ClassReplacerRector.php#L90)
> Line 90 in [6ba0b32](/rectorphp/rector/commit/6ba0b32f60f3326035f071b7e2a81d34d36ad431)
>          return new FullyQualified($newName); 
> That's the one that change class names.
> 
> If so, what is in `$name` and `$newName` variables?

It's not. When I put a die() in here, rector is executed and I get the same result

**Comment by TomasVotruba** [comment]
Oh. It might not be related to `ClassReplacerRector` at all.

Could you try running that file through another level, e.g `symfony42`?

**Comment by orklah** [comment]
> Oh. It might not be related to `ClassReplacerRector` at all.
> 
> Could you try running that file through another level, e.g `symfony42`?

Weird result on that one:

`
PHP Fatal error:  Uncaught Rector\Exception\Configuration\InvalidConfigurationException: There must be array argument in "Rector\Php\Rector\FunctionLike\ReturnTypeDeclarationRector" constructor or explicit $argument name in configuration:

Symfony\Component\Form\AbstractTypeExtension:
    getExtendedTypes: iterable
 in C:\***\vendor\rector\rector\src\DependencyInjection\Loader\RectorServiceParametersShifter.php:79
Stack trace:
#0 C:\***\vendor\rector\rector\src\DependencyInjection\Loader\RectorServiceParametersShifter.php(52): Rector\DependencyInjection\Loader\RectorServiceParametersShifter->processServices(Array)
#1 C:\***\vendor\rector\rector\src\DependencyInjection\Loader\TolerantRectorYamlFileLoader.php(50): Rector\DependencyInjection\Loader\RectorServiceParametersShifter->process(Array)
#2 C:\***\vendor\symplify\package-builder\src\Yaml\FileLoader\AbstractParameterMergin in C:\***\vendor\rector\rector\src\DependencyInjection\Loader\RectorServiceParametersShifter.php on lin
e 79
`

Maybe this will help

**Comment by TomasVotruba** [comment]
That's different bug. Will fix that.

Could you try another level until it passes?

**Comment by orklah** [comment]
No error in "Symfony41"

**Comment by TomasVotruba** [comment]
I see. I have no idea where the issue might be. But in the end, you still have to use coding standard tool, since any correct replacement will end up with with fully-qualitifed name as in this case.

**Comment by orklah** [comment]
Ok. Could it be possible to have an option to separates changes made by each rector rule? As far as I know, you can only see the file with all the rectro applied at once.

Thanks

**Comment by TomasVotruba** [comment]
I though about it, but most people use all sets in once or just few custom rules in configs. Same way you'd use coding standards.

So if you need that, have to create `rector.yml` and put rule there.
 
```yaml
services:
    SomeRector: ~
```

**Comment by TomasVotruba** [comment]
Closing original issue as *wontfix*. 

Following issues will be solved in #1036

- The symfony42
- fatal errors with duplicated use import
- replace doc type as well

**Comment by TomasVotruba** [comment]
Re-opening for the issues in previous comment. It requires more work than I expected, so we have track of it.
